### PR TITLE
Support for hardlinks with the restore command

### LINKED ISSUE #152: Support hard links
At the moment, restic saves enough meta data to restore hard links, but this is not implemented yet.

### PR DESCRIPTION
**Pull request by jgfrm**
Requires repository change to store the device a files is stored

Closes #152

### CODE REVIEW

**Review by jgfrm — COMMENTED** [review]
changed exports and comments

**Comment by jgfrm** [comment]
export datatypes

**Review by jgfrm — COMMENTED** [review]
export datatypes

**Comment by fd0** [comment]
FYI: The tests on Travis pass, it fails because some code is not properly formatted with `gofmt`: https://travis-ci.org/restic/restic/jobs/196162594#L254

I'd suggest installing the precommit-hook as outlined here: https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches
It will prevent you from committing anything that is not properly formatted, even before you push your changes to GitHub

File: src/restic/fs/file.go [hunk 1]
```diff
@@ -102,6 +102,12 @@ func Symlink(oldname, newname string) error {
 	return os.Symlink(fixpath(oldname), fixpath(newname))
 }
 
+// Hardlink creates newname as a hard link to oldname.
+// If there is an error, it will be of type *LinkError.
+func Hardlink(oldname, newname string) error {
```
Comment (fd0) [inline src/restic/fs/file.go]: Why is this function called `Hardlink()`, an not just `Link()` as the corresponding function in package `os`?

File: src/restic/hardlinks_index.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+// HardlinkData contains to number of links for an inode, and the filename
+type HardlinkData struct {
+    Links uint64
+    Name string
+}
+
+var (
+	hardLinkIndex      = make(map[HardlinkKey]*HardlinkData)
```
Comment (fd0) [inline src/restic/hardlinks_index.go]: Having just on global instance is a bad idea, make this a separate type and embed the mutex. Also, please just use a simple `Mutex`, not a `sync.RWMutex`, that's not really needed here.

File: src/restic/hardlinks_index.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	hardLinkIndexMutex = sync.RWMutex{}
+)
+
+// ExistsLink checks wether the link already exist in the index
+func ExistsLink(inode uint64, device uint64) bool {
```
Comment (fd0) [inline src/restic/hardlinks_index.go]: All functions in this file needed to be attached to the new type.

File: src/restic/node.go [hunk 4]
```diff
@@ -482,6 +500,7 @@ func (node *Node) fillExtra(path string, fi os.FileInfo) error {
 	case "file":
 		node.Size = uint64(stat.size())
 		node.Links = uint64(stat.nlink())
+		node.Device = uint64(stat.dev())
```
Comment (fd0) [inline src/restic/node.go]: Ok, that was the missing part for being able to restore hardlinks.

**Review by fd0 — CHANGES_REQUESTED** [review]
Thanks a lot for your work!

I've read through your code. I think we can do with less code. First, please use a standard `sync.Mutex`, not a `sync.RWMutex`, the latter is way more complicated to reason about and not really necessary here.

Please make the hardlink index a new type and do not just use a single global instance.

I think we do not need to decrement the link count, or even keep track of the link count during restore. There are plenty of reasons why a counterpart of a restored file was not restored (e.g. a filter), so it's impossible to reason why a hardlink could not be restored, so I think we should not print anything.

In order for this to be merged please add tests for the new type.

These are my comments so far. If you have any questions, feel free to ask, I'm happy to assist!

**Comment by jgfrm** [comment]
Thanks for the feedback.
I have a few questions:
1. "Having just on global instance is a bad idea, make this a separate type". Can you pinpoint me to an example in the source I can have a look at? I need to create the link index on beforehand, and I thought a global instance is the way to do that.
2. "All functions in this file needed to be attached to the new type". Some questions: Is there an example in the code I can look at?

Thanks,

JG

**Comment by fd0** [comment]
Hey, you can have a look at the `Lock` struct for example, defined here: https://github.com/restic/restic/blob/master/src/restic/lock.go#L26, here's and example for a method of that type: https://github.com/restic/restic/blob/master/src/restic/lock.go#L118

**Comment by jgfrm** [comment]
I had a look at the source, but I do not precisely understand the issue.
Just started with Go

+var (
+	hardLinkIndex      = make(map[HardlinkKey]*HardlinkData)
Having just on global instance is a bad idea, make this a separate type and embed the mutex.

**Comment by fd0** [comment]
I'd suggest going through the excellent [Go Tour](https://tour.golang.org/) if you haven't done already, especially the section on ["more types"](https://tour.golang.org/moretypes/2) and ["methods"](https://tour.golang.org/methods/1).

As for a concrete proposal, make a new type like this:
```go
type HardlinkIndex struct {
    m sync.Mutex
    Index map[HardlinkKey]string // holds [inode,device] -> filename
}
```

and attach the functions to the new type, e.g. a `Has()` method:
```go
func (idx *HardlinkIndex) Has(inode, device uint64) bool {
    idx.m.Lock()
    defer idx.m.Unlock()
    _, ok := idx.Index[HardlinkKey{Inode: inode, Device: device}]
    return ok
}
```

Let me know if you need more assistance. And thanks again for jumping into the cold water and trying this yourself!

**Comment by jgfrm** [comment]
Thanks for your help.
The problem is however that the index shoudl be created once for the full restore process.
So, the node cannot create the index.
This was the motivation to use a global variable for the index

**Comment by jgfrm** [comment]
Hi,

I have xattrs for files working.
There are however two problems.

1.	I need a full path to the file (in case the node is for a file which is there) and a full path to the directory (in case the node is for a dir).
What happens is that fileWorker uses a full path:
node, err := restic.NodeFromFileInfo(e.Fullpath(), e.Info())
where dirWorker uses a relative path
n, err := restic.NodeFromFileInfo(dir.Path(), dir.Info())

My question:

*	how can I obtain the full path of a dir, specifically in node.fillExtra

2.	I also tried to implement xattr for fuse.
To call the proper syscall of the fuse package (syscallx.Setxattr), I need also the full path of the file or dir.
Again here: how can I obtain the full path?

Thanks

**Comment by fd0** [comment]
So, then I'd suggest adding a field for the index to [the Restorer struct](https://github.com/restic/restic/blob/master/src/restic/restorer.go#L14-L16) and an additional parameter to [CreateAt()](https://github.com/restic/restic/blob/master/src/restic/node.go#L100) for the index.

**Comment by fd0** [comment]
This PR is superseded by #763, closing.

**Pull request closed without merging**